File: Scripts/flumut_position_map.py
#!/usr/bin/env python3
"""flumut_position_map.py

Emit the mapping from FluMut's marker numbering onto THIS run's reference
coordinates, so downstream readers join markers to variants by lookup instead of
inferring an offset.

Why this exists
---------------
A FluMut marker is written `HA1-5:G224S` -- protein HA1, residue 224, in the
numbering of FluMut's own reference. Nothing in the marker string says where
residue 224 falls in the reference a given run was mapped against. Downstream
this was being recovered by searching for a constant shift that put every
marker's residue onto the translated reference, which fails in two ways:

  * it needs several markers per protein to pin a shift down, so small proteins
    (M2, PB1-F2, NS-2) never resolve at all; and
  * a constant shift is only correct when the two proteins are colinear. HA and
    NA have diverged by INDELS between subtypes, so no single offset is right
    across the protein -- on an H3N2 run the best shift for HA1 explained 8 of
    14 markers, which is neither a fit nor a clean failure.

Both problems disappear here. FluMut ships its reference sequences and their CDS
annotations in flumut_db.sqlite, so the mapping is a fact to be read and aligned,
not a parameter to be estimated. Aligning protein-to-protein handles the indels,
which is what makes HA and NA work across subtypes rather than only on H5N1.

Output: flumut_position_map.tsv, one row per FluMut residue that aligns.

    label       marker prefix as it appears in results (`HA1-5`, `NS-1`, `PB2`)
    protein     FluMut protein the label belongs to, per the database
    flumut_pos  residue number as written in the marker
    flumut_aa   residue FluMut's reference carries there
    product     product name in THIS run's GTF
    ref_pos     codon number in THIS run's CDS  <- the join key
    ref_aa      residue THIS run's reference carries there
    identical   whether the two references agree at that residue

A position absent from the file did not align and must not be placed. That is
deliberate: an absent position is recoverable, a confidently wrong one is not.
"""
import argparse
import os
import re
import sqlite3
import sys

CODONS = {}
_B, _AA = 'TCAG', 'FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG'
for _k, (_i, _j, _l) in enumerate((i, j, l) for i in _B for j in _B for l in _B):
    CODONS[_i + _j + _l] = _AA[_k]


def translate(cds):
    return ''.join(CODONS.get(cds[i:i + 3].upper(), 'X') for i in range(0, len(cds) - 2, 3))


def norm_gene(g):
    """Strip a trailing subtype suffix only: HA_H3 -> HA, NA_N2 -> NA.

    Must leave PA-X, PB1-F2, NS1, M1 and M2 untouched -- same rule as makeGTF.R
    and the viewer, so all three agree on what a product is called.
    """
    return re.sub(r'_[HN]?\d+$', '', str(g))


def short_locus(x):
    return re.sub(r'_[A-Z][0-9]+$', '', re.sub(r'^A_', '', str(x or '')))


# FluMut protein -> product name in our GTF. FluMut splits HA into its two mature
# subunits and numbers each from 1; our GTF has one HA ORF, so both map onto it
# and the alignment is what places them.
PRODUCT = {'HA1': 'HA', 'HA2': 'HA', 'NA': 'NA', 'NS-1': 'NS1', 'NS-2': 'NEP',
           'M1': 'M1', 'M2': 'M2', 'NP': 'NP', 'PA': 'PA', 'PA-X': 'PA-X',
           'PB1': 'PB1', 'PB1-F2': 'PB1-F2', 'PB2': 'PB2'}


def find_db(explicit):
    if explicit:
        return explicit
    try:
        import flumutdb
        p = os.path.join(os.path.dirname(flumutdb.__file__), 'flumut_db.sqlite')
        if os.path.exists(p):
            return p
    except ImportError:
        pass
    return None


def read_fasta(path):
    seqs, nm, buf = {}, None, []
    with open(path) as fh:
        for line in fh:
            line = line.strip()
            if not line:
                continue
            if line[0] == '>':
                if nm:
                    seqs[nm] = ''.join(buf)
                nm, buf = line[1:].split()[0], []
            else:
                buf.append(line)
    if nm:
        seqs[nm] = ''.join(buf)
    return seqs


def read_our_proteins(reference_fa, gtf_dir):
    """Translate this run's CDS exactly as makeGTF.R defines them.

    combined.gtf is skipped and identical intervals are de-duplicated, because a
    repeated CDS record would otherwise double a product's length.
    """
    cds, seen = {}, set()
    for fn in sorted(f for f in os.listdir(gtf_dir) if f.endswith('.gtf')):
        if fn.endswith('combined.gtf'):
            continue
        with open(os.path.join(gtf_dir, fn)) as fh:
            for line in fh:
                if not line.strip() or line[0] == '#':
                    continue
                c = line.rstrip('\n').split('\t')
                if len(c) < 9 or c[2] != 'CDS':
                    continue
                m = re.search(r'gene_id "([^"]+)"', c[8])
                if not m:
                    continue
                g = norm_gene(m.group(1))
                key = (g, c[0], c[3], c[4])
                if key in seen:
                    continue
                seen.add(key)
                cds.setdefault(g, {'seqname': c[0], 'exons': []})['exons'].append(
                    (int(c[3]), int(c[4])))

    seqs = read_fasta(reference_fa)
    prot = {}
    for g, rec in cds.items():
        rec['exons'].sort()
        seq = seqs.get(rec['seqname'])
        if seq is None:
            for k, v in seqs.items():
                if short_locus(k) == short_locus(rec['seqname']):
                    seq = v
                    break
        if seq is None:
            continue
        prot[g] = translate(''.join(seq[s - 1:e] for s, e in rec['exons']))
    return prot


def read_flumut_proteins(db):
    """Splice and translate each protein from FluMut's own reference sequences."""
    con = sqlite3.connect(db)
    refs = {name: seq for name, seq in con.execute('select name, sequence from "references"')}
    ann = {}
    for start, end, protein, refname in con.execute(
            'select start, end, protein_name, reference_name from annotations'):
        ann.setdefault(protein, {'ref': refname, 'exons': []})['exons'].append((start, end))

    labels = {}
    for name, protein in con.execute('select name, protein_name from mutations'):
        labels.setdefault(name.split(':')[0], protein)

    version = None
    try:
        version = '.'.join(str(x) for x in list(
            con.execute('select major, minor from db_version'))[0])
    except (sqlite3.Error, IndexError):
        pass
    con.close()

    prot = {}
    for protein, rec in ann.items():
        seq = refs.get(rec['ref'])
        if seq is None:
            continue
        rec['exons'].sort()
        prot[protein] = translate(''.join(seq[s - 1:e] for s, e in rec['exons']))
    return prot, labels, version


def align(our_seq, fm_seq):
    """Align FluMut's protein onto ours, returning [(our_idx0, fm_idx0), ...].

    End gaps on OUR sequence are free: HA1 and HA2 are subunits of a protein our
    GTF holds as one ORF, so the query has to be free to sit inside the target
    without paying for the flanks. Internal gaps are scored normally -- those are
    the real indels between subtypes, and they are the whole reason this is an
    alignment rather than an offset.
    """
    from Bio import Align
    from Bio.Align import substitution_matrices

    aligner = Align.PairwiseAligner()
    aligner.substitution_matrix = substitution_matrices.load('BLOSUM62')
    aligner.open_gap_score = -11
    aligner.extend_gap_score = -1
    aligner.mode = 'global'
    # Free end gaps on OUR sequence (Biopython >= 1.86 spelling; the pinned image
    # supplies it). Assign, never probe: READING this attribute raises ValueError
    # whenever the open and extend end-gap scores differ, as they do here, so
    # hasattr() would throw before aligning anything.
    aligner.end_insertion_score = 0.0

    best = aligner.align(our_seq, fm_seq)[0]
    pairs = []
    for (t0, t1), (q0, q1) in zip(*best.aligned):
        for k in range(t1 - t0):
            pairs.append((t0 + k, q0 + k))
    return pairs, best.score


def main():
    ap = argparse.ArgumentParser(description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument('--reference', required=True, help='this run\'s reference FASTA')
    ap.add_argument('--gtf', required=True, help='reference_gtf directory')
    ap.add_argument('--db', default=None, help='flumut_db.sqlite (default: from flumutdb)')
    ap.add_argument('--out', default='flumut_position_map.tsv')
    args = ap.parse_args()

    db = find_db(args.db)
    if not db:
        sys.stderr.write('flumut_position_map: flumut database not found; '
                         'no position map written\n')
        return 0

    our = read_our_proteins(args.reference, args.gtf)
    fm, labels, dbver = read_flumut_proteins(db)
    if not our:
        sys.stderr.write('flumut_position_map: no reference proteins could be translated; '
                         'no position map written\n')
        return 0

    rows, summary = [], []
    for label, protein in sorted(labels.items()):
        product = PRODUCT.get(protein)
        fm_seq = fm.get(protein)
        our_seq = our.get(product) if product else None
        if not fm_seq or not our_seq:
            summary.append(f'  {label:<8} {protein:<8} -> '
                           f'{"no FluMut protein" if not fm_seq else f"no {product} in this reference"}')
            continue

        pairs, _ = align(our_seq, fm_seq)
        ident = sum(1 for t, q in pairs if our_seq[t] == fm_seq[q])
        for t, q in pairs:
            rows.append((label, protein, q + 1, fm_seq[q], product, t + 1, our_seq[t],
                         'yes' if our_seq[t] == fm_seq[q] else 'no'))

        offsets = {t - q for t, q in pairs}
        shape = (f'offset {list(offsets)[0]:+d}' if len(offsets) == 1
                 else f'{len(offsets)} blocks (indels)')
        summary.append(f'  {label:<8} {protein:<8} -> {product:<7} '
                       f'{len(pairs):>4}/{len(fm_seq)} aligned, '
                       f'{100 * ident / max(len(pairs), 1):5.1f}% identical, {shape}')

    with open(args.out, 'w') as fh:
        fh.write('label\tprotein\tflumut_pos\tflumut_aa\tproduct\tref_pos\tref_aa\tidentical\n')
        for r in rows:
            fh.write('\t'.join(str(x) for x in r) + '\n')

    sys.stderr.write(f'flumut_position_map: flumut db {dbver or "unknown"}, '
                     f'{len(rows)} positions across {len(set(r[0] for r in rows))} labels\n')
    sys.stderr.write('\n'.join(summary) + '\n')
    return 0


if __name__ == '__main__':
    sys.exit(main())

File: Scripts/flumut_position_map.R
#!/usr/bin/env Rscript

#### flumut_position_map.R
####
#### Map FluMut's marker numbering onto this run's reference coordinates, so a
#### reader joins markers to variants by lookup rather than by an inferred offset.
####
#### A FluMut marker is written like `HA1-5:G224S`: protein HA1, residue 224, in
#### the numbering of FluMut's own reference. The marker string does not say where
#### residue 224 falls in the reference a run was mapped against. A constant shift
#### does not solve this. Small proteins carry too few markers to estimate a
#### shift, and one shift is correct only when the two proteins are colinear. HA
#### and NA differ by indels between subtypes, so no single offset holds across the
#### protein.
####
#### FluMut ships its reference sequences and their CDS annotations in
#### flumut_db.sqlite. The mapping is therefore read from the database and aligned,
#### not estimated. Protein-to-protein alignment handles the indels, which is why
#### HA and NA place across subtypes and not only on H5N1.
####
#### Alignment note: this uses a semi-global protein alignment (BLOSUM62) with free
#### end gaps on THIS run's ORF (Biostrings "global-local"), so a FluMut subunit
#### such as HA1 or HA2 sits inside the ORF without paying for the flanks. Internal
#### gaps are scored normally, because those are the real indels between subtypes.
####
#### Output: flumut_position_map.tsv, one row per FluMut residue that aligns.
####   label       marker prefix as it appears in results (HA1-5, NS-1, PB2)
####   protein     FluMut protein the label belongs to, per the database
####   flumut_pos  residue number as written in the marker
####   flumut_aa   residue FluMut's reference carries there
####   product     product name in THIS run's GTF
####   ref_pos     codon number in THIS run's CDS  (the join key)
####   ref_aa      residue THIS run's reference carries there
####   identical   whether the two references agree at that residue
####
#### A position absent from the file did not align and must not be placed.
####
#### usage:
####   Rscript flumut_position_map.R --reference ref.fa --gtf reference_gtf \
####     [--db flumut_db.sqlite] [--out flumut_position_map.tsv]

suppressMessages(library(Biostrings))
suppressMessages(library(DBI))

# Standard genetic code, built in the same T,C,A,G order the pipeline uses.
CODON.BASES = c("T", "C", "A", "G")
CODON.AA = strsplit(paste0("FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSS",
                           "RRVVVVAAAADDEEGGGG"), "")[[1]]
CODON.NAMES = character(64)
.k = 1L
for (.i in CODON.BASES) for (.j in CODON.BASES) for (.l in CODON.BASES) {
  CODON.NAMES[.k] = paste0(.i, .j, .l); .k = .k + 1L
}
CODONS = setNames(CODON.AA, CODON.NAMES)

# BLOSUM62, the same substitution matrix FluMut's own aligner uses.
data("BLOSUM62", package = "Biostrings")

# Map each FluMut protein to its product name in the GTF. FluMut splits HA into
# two mature subunits and numbers each from 1; the GTF has one HA ORF, so both
# subunits map to it and alignment determines their positions.
PRODUCT = c(HA1 = "HA", HA2 = "HA", "NA" = "NA", "NS-1" = "NS1", "NS-2" = "NEP",
            M1 = "M1", M2 = "M2", NP = "NP", PA = "PA", "PA-X" = "PA-X",
            PB1 = "PB1", "PB1-F2" = "PB1-F2", PB2 = "PB2")

# Translate a spliced CDS to protein, codon by codon, dropping a trailing partial
# codon. An unknown codon becomes X, matching FluMut's own translation.
translate.cds = function(cds) {
  cds = toupper(cds)
  n = nchar(cds)
  if (n < 3L) return("")
  starts = seq.int(1L, n - 2L, by = 3L)
  aa = vapply(starts, function(i) {
    a = unname(CODONS[substr(cds, i, i + 2L)])
    if (is.na(a)) "X" else a
  }, character(1))
  paste0(aa, collapse = "")
}

# Strip a trailing subtype suffix only: HA_H3 -> HA, NA_N2 -> NA. Leaves PA-X,
# PB1-F2, NS1, M1 and M2 untouched, matching makeGTF.R and the viewer.
norm.gene = function(g) sub("_[HN]?[0-9]+$", "", as.character(g))

short.locus = function(x) {
  x = if (length(x) == 0L || is.na(x)) "" else as.character(x)
  sub("_[A-Z][0-9]+$", "", sub("^A_", "", x))
}

# Read a FASTA into a named character vector: header first token -> sequence.
read.fasta = function(path) {
  lines = readLines(path)
  seqs = list()
  nm = NULL
  buf = character(0)
  for (line in lines) {
    line = trimws(line)
    if (!nzchar(line)) next
    if (substr(line, 1L, 1L) == ">") {
      if (!is.null(nm)) seqs[[nm]] = paste0(buf, collapse = "")
      nm = strsplit(substr(line, 2L, nchar(line)), "[[:space:]]")[[1]][1]
      buf = character(0)
    } else {
      buf = c(buf, line)
    }
  }
  if (!is.null(nm)) seqs[[nm]] = paste0(buf, collapse = "")
  unlist(seqs)
}

get.seq = function(seqs, name) {
  if (!is.null(name) && length(name) == 1L && name %in% names(seqs))
    unname(seqs[[name]]) else NULL
}

resolve.locus = function(name, seqs) {
  if (name %in% names(seqs)) return(name)
  for (k in names(seqs)) if (short.locus(k) == short.locus(name)) return(k)
  NULL
}

# CDS intervals per product, exactly as makeGTF.R defines them. combined.gtf is
# skipped and identical intervals de-duplicated, because a repeated CDS record
# would otherwise double a product's length.
read.cds = function(gtf.dir) {
  cds = list()
  seen = character(0)
  files = sort(list.files(gtf.dir, pattern = "\\.gtf$"))
  for (fn in files) {
    if (grepl("combined\\.gtf$", fn)) next
    for (line in readLines(file.path(gtf.dir, fn))) {
      if (!nzchar(trimws(line)) || substr(line, 1L, 1L) == "#") next
      c = strsplit(line, "\t", fixed = TRUE)[[1]]
      if (length(c) < 9L || c[3] != "CDS") next
      m = regmatches(c[9], regexec("gene_id \"([^\"]+)\"", c[9]))[[1]]
      if (length(m) == 0L) next
      g = norm.gene(m[2])
      key = paste(g, c[1], c[4], c[5], sep = "\t")
      if (key %in% seen) next
      seen = c(seen, key)
      if (is.null(cds[[g]])) cds[[g]] = list(seqname = c[1], exons = list())
      cds[[g]]$exons[[length(cds[[g]]$exons) + 1L]] =
        c(as.integer(c[4]), as.integer(c[5]))
    }
  }
  for (g in names(cds)) {
    ex = cds[[g]]$exons
    starts = vapply(ex, function(e) e[1], integer(1))
    cds[[g]]$exons = ex[order(starts)]
  }
  cds
}

# Translate this run's CDS exactly as makeGTF.R defines them, returning a named
# vector product -> protein.
read.our.proteins = function(reference.fa, gtf.dir) {
  cds = read.cds(gtf.dir)
  seqs = read.fasta(reference.fa)
  prot = character(0)
  for (g in names(cds)) {
    rec = cds[[g]]
    locus = resolve.locus(rec$seqname, seqs)
    if (is.null(locus)) next
    seq = get.seq(seqs, locus)
    nt = character(0)
    for (e in rec$exons) nt = c(nt, substr(seq, e[1], e[2]))
    prot[g] = translate.cds(paste0(nt, collapse = ""))
  }
  prot
}

# Splice and translate each protein from FluMut's own reference sequences.
# Returns the proteins, the marker labels (prefix -> protein), and the db version.
read.flumut.proteins = function(db) {
  con = dbConnect(RSQLite::SQLite(), dbname = db)
  on.exit(dbDisconnect(con), add = TRUE)

  ref.tab = dbGetQuery(con, 'select name, sequence from "references"')
  refs = setNames(ref.tab$sequence, ref.tab$name)

  ann.tab = dbGetQuery(con,
    "select start, end, protein_name, reference_name from annotations")
  ann = list()
  for (i in seq_len(nrow(ann.tab))) {
    p = ann.tab$protein_name[i]
    if (is.null(ann[[p]])) ann[[p]] = list(ref = ann.tab$reference_name[i],
                                           exons = list())
    ann[[p]]$exons[[length(ann[[p]]$exons) + 1L]] =
      c(ann.tab$start[i], ann.tab$end[i])
  }

  mut.tab = dbGetQuery(con, "select name, protein_name from mutations")
  labels = list()
  for (i in seq_len(nrow(mut.tab))) {
    key = strsplit(mut.tab$name[i], ":", fixed = TRUE)[[1]][1]
    if (is.null(labels[[key]])) labels[[key]] = mut.tab$protein_name[i]
  }

  version = tryCatch({
    v = dbGetQuery(con, "select major, minor from db_version")
    if (nrow(v) >= 1L) paste(v$major[1], v$minor[1], sep = ".") else NULL
  }, error = function(e) NULL)

  prot = character(0)
  for (p in names(ann)) {
    seq = if (ann[[p]]$ref %in% names(refs)) unname(refs[[ann[[p]]$ref]]) else NULL
    if (is.null(seq)) next
    ex = ann[[p]]$exons
    starts = vapply(ex, function(e) e[1], integer(1))
    ex = ex[order(starts)]
    nt = character(0)
    for (e in ex) nt = c(nt, substr(seq, e[1], e[2]))
    prot[p] = translate.cds(paste0(nt, collapse = ""))
  }
  list(prot = prot, labels = labels, version = version)
}

# Align FluMut's protein onto ours, returning a data frame of aligned index pairs
# (our_idx0, fm_idx0). FluMut's protein is the pattern and must align in full; our
# ORF is the subject and carries free end gaps ("global-local"), so a subunit sits
# inside the ORF without paying for the flanks. Gap scores reproduce the original
# aligner (open -11, extend -1): Biostrings charges gapOpening + gapExtension per
# gap position, so gapOpening 10 and gapExtension 1 give the same affine cost.
align = function(our.seq, fm.seq) {
  a = pairwiseAlignment(pattern = AAString(fm.seq), subject = AAString(our.seq),
                        type = "global-local", substitutionMatrix = BLOSUM62,
                        gapOpening = 10, gapExtension = 1)
  patc = strsplit(as.character(alignedPattern(a)), "", fixed = TRUE)[[1]]
  subc = strsplit(as.character(alignedSubject(a)), "", fixed = TRUE)[[1]]
  our.idx = integer(0)
  fm.idx = integer(0)
  ri = start(subject(a)) - 1L   # our index, 0-based
  qi = start(pattern(a)) - 1L   # fm index, 0-based
  for (j in seq_along(subc)) {
    sg = subc[j] == "-"
    pg = patc[j] == "-"
    if (!sg && !pg) {
      our.idx = c(our.idx, ri)
      fm.idx = c(fm.idx, qi)
    }
    if (!sg) ri = ri + 1L
    if (!pg) qi = qi + 1L
  }
  data.frame(our = our.idx, fm = fm.idx)
}

# Read a `--flag value` pair from the argument vector.
args = commandArgs(trailingOnly = TRUE)
get.flag = function(name, default = NULL) {
  hit = which(args == name)
  if (length(hit) == 0L) return(default)
  i = hit[length(hit)]
  if (i == length(args)) {
    cat("Error: ", name, " needs a value\n", sep = "", file = stderr())
    quit(status = 1)
  }
  args[i + 1L]
}

# Locate flumut_db.sqlite the same way FluMut does: an explicit path, else the
# copy shipped inside the flumutdb Python package on this environment's PATH.
find.db = function(explicit) {
  if (!is.null(explicit)) return(explicit)
  py = Sys.which("python3")
  if (nzchar(py)) {
    code = paste0("import flumutdb, os; ",
                  "print(os.path.join(os.path.dirname(flumutdb.__file__), ",
                  "'flumut_db.sqlite'))")
    out = suppressWarnings(system2(py, c("-c", shQuote(code)),
                                   stdout = TRUE, stderr = FALSE))
    if (length(out) >= 1L && nzchar(out[1]) && file.exists(out[1])) return(out[1])
  }
  NULL
}

reference.path = get.flag("--reference")
gtf.dir = get.flag("--gtf")
db.path = get.flag("--db")
out.path = get.flag("--out", "flumut_position_map.tsv")

if (is.null(reference.path) || is.null(gtf.dir)) {
  cat("Usage: Rscript flumut_position_map.R --reference <fa> --gtf <dir> ",
      "[--db <flumut_db.sqlite>] [--out ...]\n", sep = "", file = stderr())
  quit(status = 1)
}

db = find.db(db.path)
if (is.null(db)) {
  cat("flumut_position_map: flumut database not found; no position map written\n",
      file = stderr())
  quit(status = 0)
}

our = read.our.proteins(reference.path, gtf.dir)
fm.data = read.flumut.proteins(db)
fm = fm.data$prot
labels = fm.data$labels
dbver = fm.data$version

if (length(our) == 0L) {
  cat("flumut_position_map: no reference proteins could be translated; ",
      "no position map written\n", sep = "", file = stderr())
  quit(status = 0)
}

rows = character(0)
summary = character(0)
for (label in sort(names(labels))) {
  protein = labels[[label]]
  product = if (protein %in% names(PRODUCT)) unname(PRODUCT[[protein]]) else NULL
  fm.seq = if (protein %in% names(fm)) unname(fm[[protein]]) else NULL
  our.seq = if (!is.null(product) && product %in% names(our)) unname(our[[product]]) else NULL

  if (is.null(fm.seq) || !nzchar(fm.seq) || is.null(our.seq) || !nzchar(our.seq)) {
    msg = if (is.null(fm.seq) || !nzchar(fm.seq)) "no FluMut protein"
          else paste0("no ", product, " in this reference")
    summary = c(summary, sprintf("  %-8s %-8s -> %s", label, protein, msg))
    next
  }

  pairs = align(our.seq, fm.seq)
  our.aa = substring(our.seq, pairs$our + 1L, pairs$our + 1L)
  fm.aa = substring(fm.seq, pairs$fm + 1L, pairs$fm + 1L)
  ident = sum(our.aa == fm.aa)
  rows = c(rows, paste(label, protein, pairs$fm + 1L, fm.aa, product,
                       pairs$our + 1L, our.aa,
                       ifelse(our.aa == fm.aa, "yes", "no"), sep = "\t"))

  offsets = unique(pairs$our - pairs$fm)
  shape = if (length(offsets) == 1L) sprintf("offset %+d", offsets)
          else sprintf("%d blocks (indels)", length(offsets))
  summary = c(summary, sprintf(
    "  %-8s %-8s -> %-7s %4d/%d aligned, %5.1f%% identical, %s",
    label, protein, product, nrow(pairs), nchar(fm.seq),
    100 * ident / max(nrow(pairs), 1L), shape))
}

writeLines(c(paste("label", "protein", "flumut_pos", "flumut_aa", "product",
                   "ref_pos", "ref_aa", "identical", sep = "\t"), rows), out.path)

n.labels = if (length(rows) == 0L) 0L else length(unique(sub("\t.*$", "", rows)))
cat(sprintf("flumut_position_map: flumut db %s, %d positions across %d labels\n",
            if (is.null(dbver)) "unknown" else dbver, length(rows), n.labels),
    file = stderr())
if (length(summary) > 0L) cat(paste(summary, collapse = "\n"), "\n", sep = "",
                              file = stderr())
quit(status = 0)
